keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+workload migration. At first, we will just check deployments with a positive
+`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
+that could be checked before disrupting any workload (see below). If the cluster,
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
+compatible way.
+
+`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
+NodeMaintenance object together with marking the node as schedulable again. The default value of
+this flag will be `false`.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
+also support setting a reason for the maintenance. It is unclear whether we should allow marking the
+maintenances as finished and keeping the old maintenances in the system for bookkeeping purposes.
+
+As an alternative, it would be possible to mark the node object instead of this API. But it is
+probably better to decouple this from the node for extensibility reasons. As we can see, the
+`kubectl drain` logic is a bit complex, and it may be possible to move this logic to a controller
+in the future and make the node maintenance purely declarative. It also helps to decouple RBAC
+permissions from the node object.
+
+```go
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    NodeName string
+	Reason string
+}
+
+type NodeMaintenanceStatus struct {
+	// number of pods this node maintenance is requesting to shut down 
+	TargetedPods int32
+}
+```
+
+### NodeMaintenance Controller
+
+Node maintenance controller will be introduced and added to `kube-controller-manager`. It will
+observe NodeMaintenance objects and set `TargetedByNodeMaintenance` condition on selected pods.
+The pods would be selected according to the node. Controllers that own these pods, would observe
+them and start a removal or migration from these nodes upon detecting this condition.
+
+The selection process can be later enhanced to include priority or pod type (to target daemon sets
+which we do not currently support).
+
+The controller would also remove the `TargetedByNodeMaintenance` condition from the targeted pods
+upon premature `NodeMaintenance` object removal.
+
+The controller can then show progress by reconciling `.status.targetedPods`, to show how many pods
+remain to be removed from the node.
+
+The controller would also emit a warning event if the targeted node was not marked as unschedulable.
+We could prevent this, by stomping on any changes to the node's `.spec.unschedulable` field and set
+it to true for the whole lifetime of `NodeMaintenance` object.
+
+
+### Deployment Controller
+
+The deployment controller will watch its pods and react when it observes pods with
```

Comment:
```suggestion
The ReplicaSet controller will watch its pods and react when it observes pods with
```

Reply:
yeah, this section was very roughly written. I have updated it to include more details how it would be implemented